---
title: "mqhandleR - R Package for the python package mqhandler"
author: Klaudia Adamowicz

output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>", eval = FALSE
)
```

```{r loading, include=TRUE, warning=FALSE, message=FALSE}
  # Install mqhandleR from github and build vignettes
  if (!requireNamespace("devtools", quietly = TRUE)) install.packages("devtools")
  if(!require("mqhandleR",character.only = TRUE)) devtools::install_github("symbod/MaxQuantHandler-R", build_vignettes = TRUE, dependencies = TRUE)
  # Load and attach mqhandleR 
  library("mqhandleR")
```

## Introduction {-}

## Filter Protein IDs

```{r}
# mandatory
data = data.table::fread("<file>")
protein_column = "Protein IDs" # Name of column with protein IDs

# optional 
organism = "human" # Specify organism the ids should match to
gene_column = "Gene names" # Name of column with gene names
action = "delete" # What to do, if IDs cell is empty after filtering. Keep empty cell, delete it or fill it based on gene name.
reviewed = TRUE # Bool to indicate if newly retrieved protein IDs should be reduced to reviewed ones
decoy = FALSE # Bool to indicate if protein ids from decoy fasta (REV__, CON__) should be kept
```

```{r}
filtered_data <- filter_protein_ids(data = data, protein_column = protein_column, organism = organism,
                                    gene_column = gene_column, action = action, 
                                    reviewed = reviewed, decoy = decoy)
```

## Remap Gene Names

```{r}
# mandatory
data = data.table::fread("<file>")
mode = "<mode>" #  Mode of refilling. See below for more infos
protein_column = "Protein IDs" # Name of column with protein IDs

# optional 
organism = "human" # Specify organism the ids should match to
protein_column = "Protein IDs" # Name of column with protein IDs
gene_column = "Gene names" # Name of column with gene names
skip_filled = TRUE # Bool to indicate if already filled gene names should be skipped
fasta = "<file>" # Fasta file when mode all or fasta
```

Modes:

* all - Use primarly fasta infos and additionally uniprot infos.
* fasta - Use information extracted from fasta headers.
* uniprot - Use mapping information from uniprot and use all gene names.
* uniprot_primary - Use mapping information from uniprot and only all primary gene names.
* uniprot_one - Use mapping information from uniprot and only use most frequent single gene name.

```{r}
remapped_data <- remap_genenames(data = data, mode = mode, 
                                 protein_column = protein_column, gene_column = gene_column,
                                 skip_filled = skip_filled, organism = organism, fasta = fasta)
```

## Get Orthologs

```{r}
# mandatory
data = data.table::fread("<file>")
source_organism = "rat" # Specify organism the ids should match to
target_organism = "human" # Specify organism the ids should match to
gene_column = "Gene names" # Name of column with gene names
```

```{r}
ortholog_data <- get_orthologs(data = data, gene_column = gene_column, 
                               organism = source_organism, tar_organism = target_organism)
```